### PR TITLE
NumPy 1.16 support

### LINKED ISSUE #3741: Add support for Numpy 1.16
- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Running the test suite 

`python -m numba.runtests -m 4`

shows 77 fails under Numpy 1.16, but none under 1.15.

Many of them are related to the following:
datetime behaviour up to 1.15 

```
np.datetime64('NaT', 'Y')==np.datetime64('NaT', 'Y')
__main__:1: FutureWarning: In the future, 'NAT == x' and 'x == NAT' will always be False.
True
```
datetime behaviour under 1.16

```
np.datetime64('NaT', 'Y') == np.datetime64('NaT', 'Y')
False
```

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

Closes #3741

### CODE REVIEW

**Comment by rjenc29** [comment]
Morning, I'll have a look at refactoring `np.interp` - in the meantime, this crude feature gate will make it line up with numpy across versions, I think:
https://gist.github.com/rjenc29/3d3e70693f2fe68be1cf315d3d3a5092

I've tested locally on a couple of pre- and post-1.16 versions and it seems OK - so just thought I would flag it in case useful.

File: numba/targets/arraymath.py [hunk 1]
```diff
@@ -1352,6 +1352,28 @@ def _prepare_array_impl(arr):
...
+                return np.float64
+            dt = getattr(obj, 'dtype', None)
+            if dt is None:
+                raise TypeError("type has no dtype attr")
+            return as_dtype(dt)
+        except (TypeError, NotImplementedError) as e:
+            # has dtype, but is not something understandable
+            if 'cannot be represented' in str(e) or \
+            'data type not understood' in str(e):
+                obj = obj.dtype
```
Comment (sklam) [inline numba/targets/arraymath.py]: I think the only sensible type for `obj` is `Sequence`.  Better to explicitly check for that then to rely on exception messages.

**Comment by sklam** [comment]
The rest looks good.  Just pending CI.

**Comment by stuartarchibald** [comment]
I've merged in master to resolve conflicts and also reenabled some more skipped np 116 tests.

**Comment by stuartarchibald** [comment]
Pending CI pass I think this can be merge.

**Comment by seibert** [comment]
CI looks good.

**Pull request merged**